### DISCUSSION TITLE
Intermediate submission - deadline, instructions

### DISCUSSION THREAD

**Discussion by fpetric** [larics/icuas24_competition#27] [Announcements]
Dear teams, 

first of all, the deadline for the intermediate submission is **January 20th, 23:59 AOE.** 

## Uploading solutions 
We have sent emails to team leaders containing links to shared folders that we will use to process your solutions.

Important: If your team leader has not received a link, or you received a wrong link, let us know as soon as possible by sending email to competition email. The folder name should correspond to your team's name, without spaces and special characters (such as ,, & and similar).

## What to upload
You need to upload a .zip file named `<team_name>_docker.zip` to the appropriate folder. Again, for team name, please use the name of the folder you were shared (team name without special characters). The zip file should contain:

- your version of `icuas24_competition` package (with your session.yml and any custom config files)
- update the docker file and docker build/run scripts to enable the download and build of all your code
- all private ROS packages that you developed that cannot be downloaded using the submitted docker file

You must not include `uav_ros_simulation` package and its dependencies in your .zip file.

To increase the chance of your code working, please keep your submission as clean as possible. Never include build output from your machine in your submission.

## What should the UAV do
The UAV should spawn at the location `x=1`, `y=1` and fly through the following waypoints in the empty world:

- x=1, y=5, z=2, yaw=pi/2
- x=10, y=6, z=5, yaw=-pi/2
- x=10, y=10, z=3, yaw=0

The UAV then needs to return to the starting point with z=1, yaw=pi. Once the UAV returns to the initial point, you need to publish 42 to topic `fruit_count` (type `std_msgs/Int32`).

## What are we going to do
We are going to download your .zip file, extract it, run `build_docker` script, followed by `run_docker` script, and from inside the container we will launch everything by running `start.sh` script. Your Dockerfile, packages, session.yml file and everything else should be prepared so that the aforementioned pipeline works.

## Why?
The main goal of this submission is to debug this above pipeline for running your solutions.

**Comment by arham3001** [comment]
can you please clarify where are we supposed to perform this task in. Is it meant to be in the same world which is started when we do ./start.sh or in a completely new empty world. 
And if it is in an empty world, then can you please tell how to load the drone model in the new empty world as the model of the UAV is not available in models folder.

**Reply by cmakelabs** [reply]
Yes, this is an important question

**Reply by fpetric** [reply]
The command that spawns the UAV does so in an empty world. We then add the plant beds and plant models. To prevent them from spawning, comment out or delete these lines in session.yml:
https://github.com/larics/icuas24_competition/blob/7fbaa49585ea6114466409d6901b4c1c60efee87/startup/challenge/session.yml#L28C11-L72

**Reply by vialbruh** [reply]
not sure if only i'm facing this error but commenting out or deleting the said lines did not make any difference to my world. is there something i'm missing here?

**Reply by arham3001** [reply]
i am also facing the same error

**Reply by Ro0412** [reply]
What do you do once you change the .yml file?
do u directly just run docker again or need to catkin build again?

**Reply by vialbruh** [reply]
i did catkin build, rebuilt it and docker build as well but no use

**Reply by A4dii** [reply]
Well, when we run the ./docker_run.sh we are brought into a sim_ws environment, so i changed the session.yml file in that and it worked. Can you please confirm if this is the right approach ?

**Reply by fpetric** [reply]
You need to change the session file inside the docker container. There are multiple ways to do this, you can either edit it inside container using vim which is our default editor, install some other editor inside the container, edit the file on the host machine and then copy it to container or use vscode with docker container plugin from the host machine to edit files in the container. Once the container is built and run, any changes outside of the container have no effect on what happens inside.

**Comment by cmakelabs** [comment]
In the "What the UAV should do" section it is written that the UAV should go these position:

    x=1, y=5, z=2, yaw=pi/2
    x=10, y=6, z=5, yaw=-pi/2
    x=10, y=10, z=3, yaw=0

However, the position `x=10, y=6` will cause a collision and the drone will fly over the plant inside the bed then it will crash. Are you sure that this is what you want?

**Reply by fpetric** [reply]
The world needs to be empty. See the answer above to make that happen: https://github.com/larics/icuas24_competition/discussions/27#discussioncomment-8120709

**Comment by A4dii** [comment]
Can you please specify what is yaw, like how should it change ? Upon reaching all the desired co-ordinates or while travelling. Also please specify what topic to use for this as there are multiple.

**Reply by fpetric** [reply]
However you want it to change it is fine, as long as it has the correct value at some point in the target coordinate. To control the UAV, please see here: https://github.com/larics/icuas24_competition#controlling-the-uav

**Comment by AR-AK** [comment]


**Reply by fpetric** [reply]
`/fruit_count `(without the red namespace).

**Comment by allenthreee** [comment]
What do you mean in "_you need to publish 42 to topic fruit_count (type std_msgs/Int32_)" ?
Is **42** the number of the fruit number for the given plant variety?

**Reply by fpetric** [reply]
For this first submission, this is just a signal to our evaluation script that the UAV has finished, and the number has nothing to do with fruits since there aren't any in the empty world. Theoretically, any number should be ok, but 42 is, of course, the [Answer to the Ultimate Question of Life, the Universe, and Everything](https://simple.wikipedia.org/wiki/42_(answer)).

**Comment by cmakelabs** [comment]
Do we need to change the name of the docker image in the `docker_build.sh` and `CONTAINER_NAME` in the `docker_run.sh` file in order to match with team name or we keep it as it is `icuas24_competition` ?

**Reply by fpetric** [reply]
You can keep it as is for now.

**Comment by joaovmamute** [comment]
Greetings,

How do I register my team? Which e-mail address can I send the .zip file to? Can I try to attach it to the e-mail or can I send a link on google drive?

**Reply by fpetric** [reply]
Hi, missed this one. If you register via the link in the rulebook, we will send you the link to the folder to upload the solution.

**Comment by pulak-gautam** [comment]
I am facing this error while building the docker image - ``ERROR: failed to solve: lmark1/uav_ros_simulation:focal: docker.io/lmark1/uav_ros_simulation:focal: not found``. The docker image seems to have been removed from docker hub itself. 
I had removed the docker image cached in my local, am not able to continue work now. Could you please clarify this?

**Reply by fpetric** [reply]
@lmark1 can you check this?

**Reply by pulak-gautam** [reply]
Further, https://github.com/larics/uav_ros_stack has been made private; which is a dependency of https://github.com/larics/uav_ros_simulation; it has been listed in .gitman yaml. Could you also please verify this? 
Thanks

**Reply by fpetric** [reply]
We are aware of that. Currently, native installation of uav_ros_simulation is not working. We have packaged binaries so Docker installation should be working. 

If you update the repo and the Dockerfile, you should be pulling from lmark1/uav_ros_simulation:focal-bin-0.0.1 (or the non-gpu alternative)

**Comment by pulak-gautam** [comment]
Just one small clarification, are we required to continuously keep publishing 42 to  /fruit count, after reaching initial position? or, do we just publish one message and quit?

**Reply by fpetric** [reply]
Not really important, our script will kill the simulation after the first message received, so you do not have to quit anything. If you are sending just one message, it would be wise to latch it (in general, not just here).

**Comment by pulak-gautam** [comment]
Hi, our team uploaded solutions to the provided link. However, we misspelled the name of the .zip file. We are now not able to edit our submitted .zip file or even delete it Could you kindly clarify this. We have also sent a mail regarding this.
Thank you

**Reply by fpetric** [reply]
No worries. If you need to change something, just upload the other file.

**Comment by varun-sappa** [comment]
Hi, could you give a tentative idea as to by when we could get a notification of the results of the intermediate submission?
